### ISSUE TITLE
Design Metrics Selection panel

### ISSUE THREAD

**Issue by jimmymadon** [google/site-kit-wp#6258]
## Feature Description

Clicking on the “I’ll pick metrics myself” secondary button within the setup banner or the “Change metrics” edit button on the Key metrics widget (after the question flow) will allow an authenticated admin to select their own metrics. This metrics selection panel overlay with an accordion-style list of checkboxes will slide out listing the available metrics to choose from.

![Screenshot 2022-12-11 at 12 16 47](https://user-images.githubusercontent.com/1721319/206902980-abdff1a3-2baa-43bd-9240-afd6f1bf674c.png)

Refer to the [relevant section](https://docs.google.com/document/d/1k8NOo2xEHUQAcgYG0TB5sUjGcPiC8SWIalEdd9G8-9g/edit#heading=h.vhwfv6130t0d) in the design doc for further context.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* When the user clicks on the "I'll pick metrics myself" secondary link within the Key Metrics Setup CTA component, a new Metrics Selection Panel should slide out as per the [Figma mock](https://www.figma.com/file/CdtrHB0HAjfGRWTnTMMQI0/Key-Metrics?node-id=1623%3A56200&t=bwOtTQ7S3Odzwt7u-0).
* The selection panel should list all the available registered widgets for the key metrics widget area (`AREA_MAIN_DASHBOARD_KEY_METRICS_PRIMARY`). Available widgets include widgets whose dependant modules are connected. These should've been correctly registered already in #6313.
* The UI title and descriptions should be copied from this [table](https://docs.google.com/document/d/1JvJ2gKxaX2PcYOnmA3-xqz4hK_ZHxj8Rd0svq6S8tiU/edit#bookmark=id.w8o73hf5pz1f) in the PRD. 
* The behaviour of opening an accordion card to show the widget description as well as hovering over/focusing/selecting the checkboxes should be as per the ["Metrics Selection Panel" flow in Figma](https://www.figma.com/file/CdtrHB0HAjfGRWTnTMMQI0/Key-Metrics?node-id=1623%3A58431&t=bwOtTQ7S3Odzwt7u-0).
* Clicking on "Save Selection" should persistently store the widget slugs in the database using the datastore action defined in #6314.
* Clicking on the "Settings" link in the footer notice should redirect the user to the settings page which will display a CTA banner in the new Key metrics settings area, asking the user to answer the questionnaire.

#### Create `MetricsSelectionPanel` component

- Create a folder named `MetricsSelectionPanel` in `assets/js/components/KeyMetrics`. In this folder, add the following files:
  - `index.js`
  - `Header.js`
  - `Footer.js`
  - `Metrics.js`

#### Add styles

- Create `assets/sass/components/key-metrics/_googlesitekit-km-selection-panel.scss`. All styles for the Metrics Selection Panel should be added to this file as we build this component/section in this issue.

#### Add necessary constants

- In `assets/js/googlesitekit/datastore/user/constants.js`:
  - Create a new constant that we are going to use as a `CORE_UI` store key to determine if the Metrics Selection Panel is opened or not, such as `KEY_METRICS_SELECTION_PANEL_OPENED_KEY` with the value `googlesitekit-key-metrics-selection-panel-opened`.
  - Create an object, say `KEY_METRICS_META`, that will contain the name and description of each metric, mapped to the metric widget slug. To do so:
    - The metric widget slugs can be found by searching for widgets in the codebase registered to the `AREA_MAIN_DASHBOARD_KEY_METRICS_PRIMARY` area.
    - The key should be the metric widget slug. Its value should be another object with keys `title` & `description` (translatable). Their values should be obtained from the relevant row of the table in the PRD linked in the ACs.
    - Do not add any widget slugs that are not actual metrics listed in the table, e.g. CTAs.

#### Add to datastore infrastructure

- In `assets/js/googlesitekit/datastore/user/key-metrics.js`:
  - Add `savedKeyMetricsSettings` property to the `baseInitialState`, and populate it during `saveKeyMetricsSettings` and `getKeyMetricsSettings` fetch reducer callback, similar to `keyMetricsSettings`.
  - Add a `resetKeyMetricsSettings` action, that sets the value of `savedKeyMetricsSettings` into `keyMetricsSettings`.
  - Add a `haveKeyMetricsSettingsChanged` selector, that compares between `savedKeyMetricsSettings` & `keyMetricsSettings` and returns `true` if there is a difference.

#### `MetricsSelectionPanel` Header
- In `assets/js/components/KeyMetrics/MetricsSelectionPanel/Header.js`:
  - Export a new component named `Header`.
  - Add a heading of `Select your metrics` (translatable), according to the Figma design.
  - Towards the right end of the heading, render an icon `<Button>` with the close icon from `assets/svg/icons/close.svg`. Clicking on the button should:
    - Dispatch the `CORE_USER` `resetKeyMetricsSettings` action.
    - Set the value of `KEY_METRICS_SELECTION_PANEL_OPENED_KEY` key in the `CORE_UI` store to `false` (using the `setValue` selector).
  - Add a sub-heading of `n of 4 metrics selected` (translatable) according to the Figma design. The `n`/number of selected metrics can be obtained by using the length of the array returned by the `getUserPickedMetrics` selector of the `CORE_USER` store.
  - The heading should be styled exactly/as close as possible as the Figma designs (typography, padding, etc.). The styles should be added to `assets/sass/components/key-metrics/_googlesitekit-km-selection-panel.scss`.

#### `MetricsSelectionPanel` Footer

- In `assets/js/components/KeyMetrics/MetricsSelectionPanel/Footer.js`:
  - Export a new component named `Footer`.
  - It should contain the footer of the Metrics Selection Panel, as shown in the Figma designs.
  - Primary CTA:
    - It should be a `<SpinnerButton>` labelled `Apply changes` (translatable).
    - It should be disabled if:
      - `CORE_USER` `haveKeyMetricsSettingsChanged` selector returns `false`.
      - `getUserPickedMetrics` selector of the `CORE_USER` store returns an array with length of less than 2, or greater than 4.
    - On click, it should dispatch the `saveKeyMetricsSettings` action of the `CORE_USER` store.
    - If the dispatch was successful without any error, set the value of `KEY_METRICS_SELECTION_PANEL_OPENED_KEY` key in the `CORE_UI` store to `false` (using the `setValue` selector).
  - Clicking on the `Cancel` (translatable) link should:
    - Dispatch the `CORE_USER` `resetKeyMetricsSettings` action.
    - Set the value of `KEY_METRICS_SELECTION_PANEL_OPENED_KEY` key in the `CORE_UI` store to `false` (using the `setValue` selector).
  - Above the actions, an `<ErrorNotice>` should be rendered:
    - If one of the following conditions are met (in order):
      - `getUserPickedMetrics` selector of the `CORE_USER` store returns an array with length of less than 2.
      - The `saveKeyMetricsSettings` action encountered an error
	- If the first condition is met, the error message should be `Select at least 2 metrics` (translatable).
	- Otherwise, if the second condition is met, the error message should be whatever the action encountered.
  - Clicking on `Settings` link in the footnote should navigate the user to `/wp-admin/admin.php?page=googlesitekit-settings#/admin-settings`.
  - The footer should be styled exactly/as close as possible as the Figma designs (typography, background, spacings, etc.). The styles should be added to `assets/sass/components/key-metrics/_googlesitekit-km-selection-panel.scss`.

#### `MetricsSelectionPanel` Metric listing

- In `assets/js/components/KeyMetrics/MetricsSelectionPanel/Metrics.js`:
  - Export a new component named `Metrics`.
  - Run the `CORE_WIDGETS` `getWidgets` selector with the `AREA_MAIN_DASHBOARD_KEY_METRICS_PRIMARY` area.
  - Filter the `KEY_METRICS_META` constant object so that it only returns the widget slugs that also exist in the above `getWidgets` selector call. That way, we get a list of the key metrics widgets that are registered, excluding non-metric widgets, e.g. CTAs.
  - For each widget, render the `Accordion` component. In the `Accordion` component:
    - Render a `Checkbox` component inside the `title` prop. In the `Checkbox` component:
      - Its label (child) should be the widget `title` obtained from the `KEY_METRICS_META` object.
      - It should be `checked` if the respective widget slug is present in the `CORE_USER` store's `keyMetricsSettings.widgetSlugs` object (obtainable by using the `CORE_USER` `getUserPickedMetrics` selector).
      - On click, it should add/remove the widget slug from the `CORE_USER` store's `keyMetricsSettings.widgetSlugs` object (using the `CORE_USER` `setKeyMetricsSetting` action).
      - If the `Checkbox` is not checked, and `getUserPickedMetrics` returns 4 widget slugs, disable the `Checkbox`.
    - The `Accordion` description (child) should be the `description` obtained from the `KEY_METRICS_META` object.
  - The `Accordion` and `Checkbox` components should be styled exactly/as close as possible as the Figma designs (outlines, hover/focus styles, etc). The styles should be added to `assets/sass/components/key-metrics/_googlesitekit-km-selection-panel.scss`.

#### `MetricsSelectionPanel` Wrapper & layout

- In `assets/js/components/KeyMetrics/MetricsSelectionPanel/index.js`:
  - Export a new component named `MetricsSelectionPanel`.
  - This component should act as a wrapper for the Metrics Selection Panel.
  - The entire component should be absolutely positioned to the right, with no width and 100% vertical height,
  - The classname of the wrapper should include an `--open` variant, which should apply when `KEY_METRICS_SELECTION_PANEL_OPENED_KEY` key from the `CORE_UI` store is `true`.
  - In viewports larger than mobile, this `--open` variant classname should apply a width of 384px (according to Figma) to the panel. Apply a smooth transition/animation to the width so that it imitates a slide-in from right when opened.
  - In mobile viewports, the panel should act like a modal that takes up the whole screen instead of a slide-over component.
  - Fix the `Footer` element to the bottom of the panel, allowing the `Metrics` component to take up the remaning space in the middle. It should have a `scroll` `overflow` so that the metrics are scrollable. The styles should be added to `assets/sass/components/key-metrics/_googlesitekit-km-selection-panel.scss`.

#### Storybook

- Add Storybook stories for the Metrics Selection Panel. There should be at least two variants:
  - One with an action to open/close the panel.
  - Another with the panel opened, including VRT scenarios.

### Test Coverage

- Add JS tests to verify:
  - modules listing according to the widget registrations.
  - inability to check more than 4 metrics.
  - inability to save when less than 2 metrics are checked.
- In `assets/js/googlesitekit/datastore/user/key-metrics.test.js`:
  - Add tests for `resetKeyMetricsSettings` and `haveKeyMetricsSettingsChanged`.

## QA Brief

- Set up Site Kit on a fresh site.
- Enable the `userInput` feature flag via the tester plugin.
- In the Key Metrics Setup CTA that appears in the SK dashboard, click on the `I'll pick metrics myself` action.
- Verify that a metrics selection panel slides-in from the right end of your screen.
- Verify that it looks and behaves according to the ACs and Figma designs.
- Verify that the heading section correctly displays the number of metrics selected.
- Verify that the metrics' titles and descriptions are according to the document table linked in the ACs.
- Verify that no metrics that are dependent on disconnected modules are listed. Currently, all metrics except `How people find your site` are dependent on GA4. So, if GA4 is not connected, only this metric should be listed.
- Verify that the panel doesn't let you select more than 4 metrics.
- Verify that the panel requires you to select at least 2 metrics in order to be able to save.
- Verify that clicking on the close icon or the `Cancel` action resets any unsaved changes, i.e. when you open the panel again, your changes should be lost.
- Verify that the `Settings` link in the footer works according to the ACs.
- Verify that selecting and saving certain metrics make them show up in the dashboard.

## Changelog entry

* Add management panel for selected key metrics.

**Comment by tofumatt** [comment]
(Thanks for putting a screenshot in the issue itself, it's so helpful when creating epics like this with lots of interlinked/dependant UI issues ❤️)

ACs are good, moving to IB 👍🏻

**Comment by tofumatt** [comment]
> - In `assets/js/components/KeyMetrics/MetricsSelectionPanel/index.js`:
>   - Export a new component named `MetricsSelectionPanel`.
>   - This component should act as a wrapper for the Metrics Selection Panel.
>   - The entire component should be absolutely positioned to the right, with no width and 100% vertical height,
>   - The classname of the wrapper should include an `--open` variant, which should apply when `KEY_METRICS_SELECTION_PANEL_OPENED_KEY` key from the `CORE_UI` store is `true`. This `--open` variant classname should apply a width of 384px (according to Figma) to the panel.

What should the behaviour here be on mobile? I would think this won't quite translate to a phone, especially one whose width is <`384px`.

I'd think for mobile viewports we should be explicit here and say that on any mobile screen, this panel acts more like a modal that takes up the whole screen than a slide-over component.

Other than that I think this is good, though for IBs like this it might be easier to read/scan to use sub-headings that just a list/nested list. I know it's a huge issue 😅 But headings might make it a bit easier to parse/separate out. 🙂

**Comment by nfmohit** [comment]
> What should the behaviour here be on mobile? I would think this won't quite translate to a phone, especially one whose width is <`384px`.
> 
> I'd think for mobile viewports we should be explicit here and say that on any mobile screen, this panel acts more like a modal that takes up the whole screen than a slide-over component.

That is an excellent point. I have updated the respective section to consider a full-screen modal-like structure for mobile viewports.

> Other than that I think this is good, though for IBs like this it might be easier to read/scan to use sub-headings that just a list/nested list. I know it's a huge issue 😅 But headings might make it a bit easier to parse/separate out. 🙂

I agree. I have added some headings for added context. I hope this is looking slightly better now.

Back to you for another pass, thank you @tofumatt!

**Comment by tofumatt** [comment]
Amazing, thanks so much @nfmohit, the IB is way more readable now too. 😲

IB ✅

**Comment by nfmohit** [comment]
Hi @bethanylang. I've moved this back to EB because it isn't added to the current sprint yet. I've assigned it to myself because I've worked on its IB, and since it is a large issue, I think I'll be able to execute it quicker. Feel free to unassign if you think otherwise. Thanks!

**Comment by bethanylang** [comment]
Got it, thanks @nfmohit!

**Comment by nfmohit** [comment]
Thank you for the heads-up, @jimmymadon. Those are great catches. I'll wait for the QA team to complete their check and open a follow-up PR addressing altogether.

**Comment by mohitwp** [comment]
### QA Update ❌ 

Thank you, @jimmymadon !

@nfmohit Please find issue and questions details below

**Issue 1>** As Jimmy mentioned above, accordion title is not changing its color to gray in disabled state.

<details>

![image](https://github.com/google/site-kit-wp/assets/94359491/f6b61af3-c665-48b8-98ec-23b325d51e68)

</details>

**Issue 2>** As Jimmy mentioned above, The accordion cannot be opened via the keyboard.

**Issue 3>** On view only dashboard, CTA component is showing and user is able to pick metrics.

**Issue 4>** Selected card focus state is not implemented. Check boxes have focus state when checked and unchecked. But card don't have any focus state. Although, It is mentioned under Figma.

<details>

![image](https://github.com/google/site-kit-wp/assets/94359491/27323b03-6bdb-4984-bbd4-d681690f7bfe)

</details>

**Question>**  I compared the name and description available under [PRD](https://docs.google.com/document/d/1JvJ2gKxaX2PcYOnmA3-xqz4hK_ZHxj8Rd0svq6S8tiU/edit#bookmark=id.w8o73hf5pz1f) doc and found that one metric is missing under selection panel i.e "Top earning content" . Is it expected and intentional ?

![image](https://github.com/google/site-kit-wp/assets/94359491/2e817e9a-8d2b-4cdb-8cd6-9b2423544302)

**Comment by mohitwp** [comment]
### QA Update ⚠️ 

- Tested on dev environment.
- Design issue reported [above](https://github.com/google/site-kit-wp/issues/6258#issuecomment-1666996003) are resolved now.
- Verified all points mentioned under QAB and AC.

@nfmohit @jimmymadon 

I have few questions related to my observations-

**Q-1)** As mentioned in QAB  all metrics except ` How people find your site` are dependent on GA4. So, if GA4 is disconnected then we are hiding those metrics and showing only 1 metrics i.e ` How people find your site` . But, user is not able to select that metrics because minimum 2 metrics selection is required to show the widgets. 

<details>

![image](https://github.com/google/site-kit-wp/assets/94359491/f57cbfdb-9607-4db3-a9d7-2f3f4c157bb5)

</details>

**Q-2)** Under panel we have option to select metric `"Most Popular products". ` But, test site don't have any kind of product. Should we display this option even if site don't have any products ?

<details>

![image](https://github.com/google/site-kit-wp/assets/94359491/140a7aa0-de73-4587-ab8f-c83aa7666da9)

</details>

**Q-3)** I've noticed that now on selection of a couple of metrics I'm getting widget which inform user that `AdSense is disconnected. Connect AdSense.`  So, as per above stated logic if GA4 is disconnected we are hiding all related metrics. But same is not happening if AdSense is disconnected. Is this expected?
 
<details>

![image](https://github.com/google/site-kit-wp/assets/94359491/23eef721-fcd1-46c6-bf8b-5e0cacdd2ed7)

</details>

**Q-4)** Previously, metrics under panel are showing in the same sequence as first appearance. But, now I noticed that selected metrics are showing first in the order. This is happening after merging of #6259. Is this expected ?


<details>

![image](https://github.com/google/site-kit-wp/assets/94359491/6822b634-ee82-4d98-b74e-5d174791edd3)

![image](https://github.com/google/site-kit-wp/assets/94359491/84b0e22e-e555-440c-98bd-b87ce3d4afe5)

![image](https://github.com/google/site-kit-wp/assets/94359491/e5c837d3-66ef-442c-a75b-b483251e0242)

</details>

**Comment by nfmohit** [comment]
Hi @mohitwp , thank you for sharing your observations!

> **Q-1)** As mentioned in QAB all metrics except ` How people find your site` are dependent on GA4. So, if GA4 is disconnected then we are hiding those metrics and showing only 1 metrics i.e ` How people find your site` . But, user is not able to select that metrics because minimum 2 metrics selection is required to show the widgets.

Ideally, a user shouldn't be able to open the metrics selection panel in such a scenario. I understand the `Change Metrics` link appears for the user, but that should be hidden. @jimmymadon Do you think a new issue is needed to address this?

> **Q-2)** Under panel we have option to select metric `"Most Popular products". ` But, test site don't have any kind of product. Should we display this option even if site don't have any products ?

I'll leave that question up for @jimmymadon, thanks!

> **Q-3)** I've noticed that now on selection of a couple of metrics I'm getting widget which inform user that `AdSense is disconnected. Connect AdSense.` So, as per above stated logic if GA4 is disconnected we are hiding all related metrics. But same is not happening if AdSense is disconnected. Is this expected?

It looks like the AdSense Top Earning Content widget registration no longer includes `adsense` as a module dependency. @jimmymadon Do you know if that is intentional? I remember it was a dependency before.

> **Q-4)** Previously, metrics under panel are showing in the same sequence as first appearance. But, now I noticed that selected metrics are showing first in the order. This is happening after merging of #6259. Is this expected ?

That is expected.

Thanks!

**Comment by techanvil** [comment]
As discussed on [Slack](https://10up.slack.com/archives/CBKKQEBR9/p1691496806439259), I'm going to create a followup to #7337 to update the Top Earning Content widget registration to be dependent on `adsense` as well as `analytics-4`, and also remove said widget from the list that's used to populate the panel.

**Comment by techanvil** [comment]
Update: That followup PR is now created: https://github.com/google/site-kit-wp/pull/7413

Once this is merged, the Top Earning Content widget should no longer be able to select, which should mitigate Q-3 above.

**Comment by techanvil** [comment]
Update: The PR mentioned above is now merged, so this issue should be revisited for QA with regard to Q-3.

**Comment by mohitwp** [comment]
### QA Update ⚠️ 


- Tested on dev environment.
- Verified all points mentioned under QAB and AC.
- Design issue reported [above](https://github.com/google/site-kit-wp/issues/6258#issuecomment-1666996003) are resolved now.
- For point 1 mentioned [above](https://github.com/google/site-kit-wp/issues/6258#issuecomment-1669236255) waiting for @jimmymadon input on Mariya's feedback. 
- For point 2) mentioned [above](https://github.com/google/site-kit-wp/issues/6258#issuecomment-1669236255) @jimmymadon created separate ticket - #7420 
- Verified point 3) is resolved by #7413 .


<details>

![image](https://github.com/google/site-kit-wp/assets/94359491/6822b634-ee82-4d98-b74e-5d174791edd3)

![image](https://github.com/google/site-kit-wp/assets/94359491/84b0e22e-e555-440c-98bd-b87ce3d4afe5)

![image](https://github.com/google/site-kit-wp/assets/94359491/e5c837d3-66ef-442c-a75b-b483251e0242)

https://github.com/google/site-kit-wp/assets/94359491/0bbfcb2e-da16-4d11-a07c-8428a6c63463


</details>

**Comment by mohitwp** [comment]
###  QA Update ✅ 

Thank you, @jimmymadon !

- Tested on dev environment.
- Verified all points mentioned under QAB and AC.
- Design issue reported [above](https://github.com/google/site-kit-wp/issues/6258#issuecomment-1666996003) are resolved now.
- For point 1 mentioned [above](https://github.com/google/site-kit-wp/issues/6258#issuecomment-1669236255) waiting for @jimmymadon created a new ticket #7425 
- For point 2) mentioned [above](https://github.com/google/site-kit-wp/issues/6258#issuecomment-1669236255) @jimmymadon created separate ticket - #7420 
- Verified point 3) is resolved by #7413.


<details>

![image](https://github.com/google/site-kit-wp/assets/94359491/6822b634-ee82-4d98-b74e-5d174791edd3)

![image](https://github.com/google/site-kit-wp/assets/94359491/84b0e22e-e555-440c-98bd-b87ce3d4afe5)

![image](https://github.com/google/site-kit-wp/assets/94359491/e5c837d3-66ef-442c-a75b-b483251e0242)

https://github.com/google/site-kit-wp/assets/94359491/0bbfcb2e-da16-4d11-a07c-8428a6c63463


</details>
